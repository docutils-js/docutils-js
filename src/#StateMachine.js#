class StateMachine {
    /* 
        Initialize a `StateMachine` object; add state objects.

       Parameters:

        - `state_classes`: a list of `State` (sub)classes.
        - `initial_state`: a string, the class name of the initial state.
        - `debug`: a boolean; produce verbose output if true (nonzero).
	*/

    constructor(stateClasses, initialState, debug) {
	if(debug === undefined) {
	    debug = false;
	}
	this.inputLines = undefined;
	this.inputOffset = 0;
	this.line = undefined;
	this.lineOffset = -1;
	this.debug = debug;
	this.initialState = initialState;
	this.currentState = initialState;
	this.states = {};
	this.addStates(stateClasses);
	this.observers = []
	//       self._stderr = ErrorOutput()
    }

    unlink() {
	Object.values(this.states).forEach(s => s.unlink());
	this.states = undefined;
    }

    run(inputLines, inputOffset = 0, context, inputSource, initialState) {
	/*
        Run the state machine on `input_lines`. Return results (a list).

        Reset `self.line_offset` and `self.current_state`. Run the
        beginning-of-file transition. Input one line at a time and check for a
        matching transition. If a match is found, call the transition method
        and possibly change the state. Store the context returned by the
        transition method to be passed on to the next transition matched.
        Accumulate the results returned by the transition methods in a list.
        Run the end-of-file transition. Finally, return the accumulated
        results.

        Parameters:

        - `input_lines`: a list of strings without newlines, or `StringList`.
        - `input_offset`: the line offset of `input_lines` from the beginning
          of the file.
        - `context`: application-specific storage.
        - `input_source`: name or path of source of `input_lines`.
        - `initial_state`: name of initial state.
	*/
	this.runtimeInit();
	//if isinstance(input_lines, StringList):
	//self.input_lines = input_lines
	//else:
	//self.input_lines = StringList(input_lines, source=input_source)
	this.inputLines = inputLines;
	this.lineOffset = -1;
	this.currentState = initialState || this.initialState;
	let transitions = undefined;
	const results = [];
	let state = this.getState();
	let nextState;
	try {
	    [ context, result ] = state.bof(context);
	    results.push(result);
	    while(true) {
		try {
		    try {
			this.nextLine();
			[ context, nextState, result ] = this.checkLine(context, state, transitions);
		    }
		    catch(error) {
		    }
		}
		catch(error) {
		}
	    }
	}
	catch(error) {
	}
	this.observers = [];
	return results;
    }

    addStates(stateClasses) {
	// stub
    }

    runtimeInit() {
	Object.values(this.states).forEach(s => s.runtimeInit());
    }

    getState(nextState) {
	if(nextState) {
	    this.currentState = nextState;
	}
	
	    
}

export default StateMachine;
