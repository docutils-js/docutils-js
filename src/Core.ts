import {Publisher} from './Publisher';
import { ConfigSettings, SettingsSpecType, LoggerType } from './types';
import Writer from './Writer';
import Parser from './Parser';
import Reader from './Reader';
import { Settings } from '../gen/Settings';
import { defaultPublisherOptions, defaultUsage, defaultDescription } from './constants';
export { Publisher };

/* We need a non command-line parsing based function */

/* We should document all the arguments to this function */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function publish(args: any): void {
    const myArgs = { ...defaultPublisherOptions, ...args };
    const {
        reader, readerName, parser, parserName, writer, writerName,
        settings,
    } = myArgs;
    const pub = new Publisher({
        reader, parser, writer, settings, logger:myArgs.logger,
    });
    pub.setComponents(readerName, parserName, writerName);
}


export interface PublishCmdLineArgs {
    reader?: Reader;
    readerName?: string;
    parser?: Parser;
    parserName?: string;
    writer?: Writer;
    writerName?: string;
    settings?: Settings;
    settingsSpec?: SettingsSpecType[];
    settingsOverrides?: ConfigSettings;
    configSection?: string;
    enableExitStatus?: boolean;
    argv?: string[];
    usage?: string;
    description?: string;
    logger: LoggerType;
}

/**
 *  Set up & run a `Publisher` for command-line-based file I/O (input and
 *  output file paths taken automatically from the command line).  Return the
 *  encoded string output also.
 *
 *  Parameters: see `publish_programmatically` for the remainder.
 *
 *  - `argv`: Command-line argument list to use instead of ``sys.argv[1:]``.
 *  - `usage`: Usage string, output if there's a problem parsing the command
 *    line.
 *  - `description`: Program description, output for the "--help" option
 *    (along with command-line option descriptions).
 *
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function publishCmdLine(args: PublishCmdLineArgs, cb: any): void {
    const _defaults = {
        readerName: 'standalone',
        parserName: 'restructuredtext',
        usage: defaultUsage,
        description: defaultDescription,
        enableExitStatus: true,
    };
    args = { ..._defaults, ...args };
    args.logger.silly('publishCmdLine');
    const {
        reader, readerName, parser, parserName, writer, writerName,
        settings, settingsSpec, settingsOverrides, configSection,
        enableExitStatus, argv, usage, description,
    } = args;
    const pub = new Publisher({
        reader, parser, writer, settings, logger:args.logger,
    });
    pub.setComponents(readerName, parserName, writerName);
    pub.publish({
        argv, usage, description, settingsSpec, settingsOverrides, configSection, enableExitStatus,
    }, cb);
}
