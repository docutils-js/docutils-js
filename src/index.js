export const __version__ = '0.14js';
