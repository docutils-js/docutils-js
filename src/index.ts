/* eslint-disable-next-line import/prefer-default-export */
export { parse } from './parse';
export { StringOutput, StringInput } from './io';
export { StandaloneReader } from './readers/standalone';
export { newDocument } from './newDocument';
import * as nodes from './nodes';
import Writer from './Writer';
import Transform from './Transform';
export { Reader } from './Reader';
export { Publisher } from './Publisher'
export { defaults } from '../gen/defaults';
export { pojoTranslate } from './fn/pojoTranslate';
export { htmlTranslate } from './fn/htmlTranslate';
export { RSTParser } from './parsers/restructuredtext';
export { XMLWriter } from './writers/xml'

/* eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars */
const __docformat__ = 'reStructuredText';

export const __version__ = '0.14js';

export {
    nodes, Writer, Transform,
};
