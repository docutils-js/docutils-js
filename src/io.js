import TransformSpec  from './TransformSpec';
import { createReadStream } from 'fs';
import ErrorOutput from './ErrorOutput';
import Input from './io/Input';
import Output from './io/Output';
import { ApplicationError } from'./Exceptions';

export class StringInput extends Input {
    constructor(...args) {
	super(...args);
	this.sourcePath = '<string>';
    }
	
    read(cb) {
	cb(undefined, this.source);
    }
}

export class StringOutput extends Output {
    constructor(...args) {
	super(...args);
	this.defaultDestinationPath = '<string>'
    }
    
    write(data) {
        //self.destination = self.encode(data) // fixme encoding
	if(Array.isArray(data)) {
	    data = JSON.stringify(data);
	}
	this.destination = data;
	return this.destination;
    }
}

export class FileInput extends Input {
    /* ew, too much logic for a constructor, with side effects etc! */
    constructor(args) {
	super(args);
	let { source, sourcePath, encoding, errorHandler, autoClose,
	      mode } = args;
	if(typeof source === 'undefined' && typeof sourcePath === 'undefined') {
	    throw new ApplicationError("FileInput: Undefined source and sourcePath");
	}
	
	if(autoClose === undefined) {
	    autoClose = true;
	}
	if(mode === undefined) {
	    mode = 'r';
	}
	
	this.autoClose = autoClose;
	this._stderr = new ErrorOutput()
	if(!source) {
	    if(sourcePath) {
		try {
		    this.source = createReadStream(sourcePath, { encoding: 'utf8' });
		}
		catch(error) {
		    console.log(error.stack);
		    throw error;
		}
	    } else {
		this.source = process.stdin;
	    }
	} else {
	    
	    // ??
	}
	if(!sourcePath) {
	    this.sourcePath = this.source.name;
	}
    }

    /* Read and decode a single file and return the data (Unicode string).
     */
    read(cb) {
	setTimeout(() => {
	let data;
	try {
	    /* reading ? */
	    if(this.source === process.stdin) {
		// do stuff
	    } else {
		data = this.source.read();
		if(data === null) {
		    console.log("read returned null ?");
		}
		cb(undefined, data);
	    }
	} catch(error) {
	    console.log(error.stack);
	    cb(error, undefined);
	}
	}, 100);
    }

    readLines() {
	return this.read().splitlines(true);
    }

    close() {
	if(this.source !== process.stdin) {
	    this.source.close();
	}
    }
}

export class FileOutput extends Output {
    
}
