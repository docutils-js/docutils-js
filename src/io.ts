import Input from './io/Input';
import Output from './io/Output';
import { ReadInputCallback, LoggerType } from "./types";

/** Direct string input. */
export class StringInput extends Input {
    public constructor(source: string, logger: LoggerType, sourcePath?: string, encoding?: string, errorHandler?: string) {
        super({source, sourcePath, encoding, errorHandler, logger});
        this.sourcePath = '<string>';
    }

    public read(cb: ReadInputCallback<string|string[]|{}>): void {
        cb(undefined, this.source);
    }
}

export class StringOutput extends Output<string> {
    public constructor(
        logger: LoggerType,
        destination?: string,
        destinationPath?: string,
        encoding?: string,
        errorHandler?: string
    ) {
        super(logger, destination, destinationPath, encoding, errorHandler);
        this.defaultDestinationPath = '<string>';

    }

    public write(data: string): string {
        // self.destination = self.encode(data) // fixme encoding
        if (Array.isArray(data)) {
            data = JSON.stringify(data);
        }
        this.destination = data;
        return this.destination;
    }
}

