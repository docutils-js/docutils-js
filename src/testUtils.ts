import { newDocument } from './newDocument';
import newReporter    from './newReporter';
import { defaultPublisherOptions } from '../src/constants';
import { getDefaultSettings } from './settingsHelper';
import winston, { format, transports, LoggerOptions, Logger } from 'winston';
import { defaultConsoleLogLevel } from './constants';
import {Publisher} from './Publisher';
import path from 'path';
import StateFactory from './parsers/rst/StateFactory';
import { LoggerType } from './types';
import process from 'process';

let logger: LoggerType|undefined;

export function createLogger(options?: LoggerOptions): Logger {
    if(logger !== undefined) {
        return logger;
    }
    let myOpt = options === undefined ? {} : { ...options};
    if(myOpt.format === undefined) {
        myOpt.format = format.json();
    }
    if(myOpt.transports === undefined) {
        myOpt.transports = [
            new transports.Console({level: defaultConsoleLogLevel}),
            new transports.File({filename: `${path.basename(process.argv[1])}-${process.pid}.log`, level: 'silly'}),
        ];
    }
    logger = winston.createLogger(myOpt);
    return logger;
}

export function createPublisher() {
    const publisher = new Publisher({ ...defaultPublisherOptions, logger: createLogger() });
    publisher.setComponents(defaultPublisherOptions.readerName, defaultPublisherOptions.parserName, 'xml');
    return publisher;
}

export function createStateMachine() {
}

export function createStateFactory(): StateFactory {
return new StateFactory({ logger: createLogger() });
}

export function createNewDocument(sourcePath: string='default') {
    return newDocument({sourcePath, logger: createLogger()}, getDefaultSettings());
}
  
export function createNewReporter(sourcePath: string='default') {
    return newReporter({sourcePath}, getDefaultSettings());
}
