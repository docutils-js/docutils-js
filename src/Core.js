import readers from './Readers';
import writers from './Writers';
import { OptionParser } from './FrontEnd';
import { Source } from './Sources';
import { FileInput, FileOutput } from './io';
import Publisher from './Publisher';

export { Publisher };

export const defaultUsage = '%prog [options] [<source> [<destination>]]';
export const defaultDescription = ('Reads from <source> (default is stdin) and writes to <destination> (default is stdout).  See <http://docutils.sf.net/docs/user/config.html> for the full reference.');


export function publishCmdLine(args, cb) {
    const _defaults = {
 readerName: 'standalone',
                        parserName: 'restructuredtext',
                        usage: defaultUsage,
                        description: defaultDescription,
                        enableExitStatus: true,
};
    args = { ..._defaults, ...args };
    const {
 reader, readerName, parser, parserName, writer, writerName,
      settings, settingsSpec, settingsOverrides, configSection,
            enableExitStatus, argv, usage, description,
} = args;
    const pub = new Publisher({
 reader, parser, writer, settings,
});
    pub.setComponents(readerName, parserName, writerName);
    pub.publish({
argv, usage, description, settingsSpec, settingsOverrides, configSection, enableExitStatus,
}, cb);
}
