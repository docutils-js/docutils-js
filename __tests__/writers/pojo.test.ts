import { Publisher } from '../../src/Core';
import { StringInput, StringOutput } from '../../src/io';
import { getDefaultSettings } from '../../src/';
import { createLogger }from '../../src/testUtils';

const currentLogLines = [];

afterEach(() => {
    if (currentLogLines.length) {
        //      console.log(currentLogLines.join('\n') + '\n');
        currentLogLines.length = 0;
    }
});

const defaultArgs = {
    readerName: 'standalone',
    parserName: 'restructuredtext',
    usage: '',
    description: '',
    enableExitStatus: true,
    writerName: 'pojo',
};

const defaultSettings = { ...getDefaultSettings()};

test('rst2pojo pipeline', () => {
    const settings = { ...defaultSettings };
    const args = { ...defaultArgs };
    const logger = createLogger();

    /* eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars */
    const debugFn = (msg: string) => {
        //      console.log(msg);
        //      currentLogLines.push(msg);
    };

    const { readerName, parserName, writerName } = args;
    const source = new StringInput(`Random test
===========
I like food.

`,
    logger);

    const destination = new StringOutput(logger);
    const pub = new Publisher({
        source, destination, settings, debug: true, debugFn,logger,
    });
    pub.setComponents(readerName, parserName, writerName);
    return new Promise((resolve, reject) => {
        pub.publish({}, (error: Error|{}|undefined) => {
            if (error) {
                reject(error);
                return;
            }
            expect(destination).toBeDefined();
            expect(destination.destination).toBeDefined();
            // @ts-ignore
            expect(JSON.parse(destination.destination)).toMatchSnapshot();
            currentLogLines.length = 0;
            resolve();
        });
    });
});
