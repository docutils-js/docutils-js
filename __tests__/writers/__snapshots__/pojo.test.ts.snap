// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rst2pojo pipeline 1`] = `
Array [
  "document",
  Object {
    "ids": Array [
      "random-te-t",
    ],
    "names": Array [
      "random test",
    ],
    "source": "<string>",
    "title": "Random test",
  },
  Array [
    Array [
      "title",
      Object {},
      Array [
        "Random test",
      ],
    ],
    Array [
      "paragraph",
      Object {},
      Array [
        "I like food.",
      ],
    ],
    Array [
      "paragraph",
      Object {},
      Array [],
    ],
  ],
]
`;
